<a href="https://colab.research.google.com/github/cnrgrl/Python/blob/master/02_The_DataFrame_and_Series_(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# uncomment the following line, if you are using google collab
!rm -r Pandas
!git clone https://github.com/Wuebbelt/Pandas.git

rm: cannot remove 'Pandas': No such file or directory
Cloning into 'Pandas'...
remote: Enumerating objects: 74, done.
remote: Counting objects: 100% (74/74), done.
remote: Compressing objects: 100% (64/64), done.
remote: Total 74 (delta 10), reused 73 (delta 9), pack-reused 0
Unpacking objects: 100% (74/74), done.


# The DataFrame and Series

The DataFrame and Series are the two primary pandas objects used throughout this book. In this chapter, we will learn how to read in data into a DataFrame and understand its components. We will also learn how to select a single column of data as a Series and understand its components.

## Reading external data with pandas

The one thing you need for data analysis is **data**. If you do not have any data, then you won't be able to use pandas to analyze it. This book contains many data sets stored externally in the `data` directory one level above where this notebook resides. Most of these datasets are stored as comma-separated value (**CSV**) files. These CSVs are human-readable and separate each individual piece of data with a comma. The comma is referred to as the **delimiter**. Despite its name, CSVs can use other one-character delimiters besides commas. 


### City of Chicago bike rides

We begin our data analysis adventure with a dataset on public bike rides from the city of Chicago. The data is contained in the `bikes.csv` file. There are about 50,000 recorded rides from 2013 through 2017. Each row of the dataset represents a single ride from a single person using the city's public bike stations. There are 19 columns of data containing information on gender, start time, trip duration, bike station name, temperature, wind speed, and more. Let's print out the first three lines of the `bikes.csv` file using Python's built-in capabilities for reading files. Take note of the commas separating each value on each line.

In [ ]:
with open('Pandas/bikes.csv') as f:
    for i in range(3):
        print(f.readline())     

trip_id,usertype,gender,starttime,stoptime,tripduration,from_station_name,latitude_start,longitude_start,dpcapacity_start,to_station_name,latitude_end,longitude_end,dpcapacity_end,temperature,visibility,wind_speed,precipitation,events

7147,Subscriber,Male,2013-06-28 19:01:00,2013-06-28 19:17:00,993,Lake Shore Dr & Monroe St,41.88105,-87.61697,11.0,Michigan Ave & Oak St,41.90096039,-87.62377664,15.0,73.9,10.0,12.7,-9999.0,mostlycloudy

7524,Subscriber,Male,2013-06-28 22:53:00,2013-06-28 23:03:00,623,Clinton St & Washington Blvd,41.883379999999995,-87.64116999999996,31.0,Wells St & Walton St,41.89993001,-87.63443007,19.0,69.1,10.0,6.9,-9999.0,partlycloudy



### Understanding the file location

Above, the string `../data/bikes.csv` was used to represent the file location of the data. This location is relative to the directory where this notebook resides on your machine. Let's cover every part of this strings to ensure we understand what it means.

The file location string begins with two dots, `..`. This translates as 'move one level above the current working directory' to the **Jupyter Notebooks** directory. Appearing next in the string is `/data`, which translates as 'move down into the `data` directory. 

Note that the forward slash was written to separate the directories. Both macOS and Linux operating systems use this forward slash to separate directories and files from one another. On the other hand, the Windows operating system uses the backslash. Fortunately, we can always use a forward slash regardless of our operating system, as Python will automatically handle the file location string for us.

The string ends with `/bikes.csv` which translates as 'reference the filename `bikes.csv`. In summary, the file location `../data/bikes.csv` represents a relative location to where the dataset resides.

### Import pandas

To use the pandas library, we need to import it into our namespace. By convention, pandas is imported and aliased to the name `pd`. After running the import statement below, we will have access to all pandas objects with `pd`. It is possible to use any other valid variable name as an alias, but it's best to use `pd` as the official documentation uses it along with most everyone else.

In [ ]:
import pandas as pd

### Reading a CSV with the `read_csv` function

pandas provides the `read_csv` function to read in CSV files into a pandas DataFrame. The first argument passed to `read_csv` needs to be the location of the file relative to the current directory as a string, which is the same string from above. Let's call this function to read in our data.

In [ ]:
bikes = pd.read_csv('Pandas/bikes.csv')

### Display DataFrame in Jupyter Notebook

We assigned the output from the `read_csv` function to the `bikes` variable name which now refers to a DataFrame object. To visually display the DataFrame, place the variable name as the last line in a code cell. By default, pandas outputs 60 rows and 20 columns. 

### `head` and `tail` methods

A very useful and simple method is `head`, which returns the first 5 rows of the DataFrame by default. This avoids long default output and is something I highly recommend when doing data analysis within a notebook. The `tail` method returns the last 5 rows by default. There will be few instances in the book where the `head` method is not used as displaying 60 rows is far too many and will take up a lot of space on a screen or page.

In [ ]:
bikes.head()

trip_id    usertype gender  ... wind_speed precipitation        events
0     7147  Subscriber   Male  ...       12.7       -9999.0  mostlycloudy
1     7524  Subscriber   Male  ...        6.9       -9999.0  partlycloudy
2    10927  Subscriber   Male  ...       16.1       -9999.0  mostlycloudy
3    12907  Subscriber   Male  ...       16.1       -9999.0  mostlycloudy
4    13168  Subscriber   Male  ...       17.3       -9999.0  partlycloudy

[5 rows x 19 columns]

The last five rows of the DataFrame may be displayed with the `tail` method.

In [ ]:
bikes.tail()

trip_id    usertype  gender  ... wind_speed precipitation        events
50084  17534938  Subscriber    Male  ...       16.1       -9999.0  partlycloudy
50085  17534969  Subscriber    Male  ...       16.1       -9999.0  partlycloudy
50086  17534972  Subscriber    Male  ...       16.1       -9999.0  partlycloudy
50087  17535645  Subscriber  Female  ...       11.5       -9999.0  partlycloudy
50088  17536246  Subscriber    Male  ...       15.0       -9999.0  partlycloudy

[5 rows x 19 columns]

### First and last `n` rows
Both the `head` and `tail` methods accept a single integer parameter `n` controlling the number of rows returned. Here, we output the first three rows.

In [ ]:
bikes.head(3)

trip_id    usertype gender  ... wind_speed precipitation        events
0     7147  Subscriber   Male  ...       12.7       -9999.0  mostlycloudy
1     7524  Subscriber   Male  ...        6.9       -9999.0  partlycloudy
2    10927  Subscriber   Male  ...       16.1       -9999.0  mostlycloudy

[3 rows x 19 columns]

## Components of a DataFrame

The DataFrame is composed of three separate components - the **columns**, the **index**, and the **data**. These terms will be used throughout the book and understanding them is vital to your ability to use pandas. Take a look at the following graphic of our `bikes` DataFrame stylized to put emphasis on each component.

![][1]

### The columns

The columns provide a **label** for each column and are always displayed in **bold** font above the data. A column is a single vertical sequence of data. In the above DataFrame, the column name `tripduration` references all the values in that column (993, 623, 1040, etc...).

The columns are also referred to as the **column names** or the **column labels** with individual values referred to as a **column name** or **column label**.

Most DataFrames, like the one above, use strings for column names, but its possible that they can be other types such as integers. The column names are not required to be unique, though having duplicate columns would be bad practice as it's vital to be able to uniquely identify each column.

### The index

The index provides a **label** for each row and is always displayed to the left of the data. A row is a single horizontal sequence of data. For instance, the index label **3** references all the values in its row (12907, Subscriber, Male, etc...)

The index is also referred to as the **index names/labels** or the **row names/labels** with the individual values referred to as a(n) **index name/label** or **row name/label**.

In the above DataFrame, the index is simply a sequence of integers beginning at 0. The values in the index are not limited to integers. Strings are a common type that are used in the index and make for more descriptive labels.

Surprisingly, values in the index are not required to be unique. In fact, all the index values can be the same. A row label does not guarantee a one-to-one mapping to one specific row.

### The data

The actual data is to the right of the index and below the columns and is always in normal font. The data is also referred to as the **values**. The data represents all the values for all the columns. It is important to note that the index and the columns are NOT part of the data. They are separate objects that act as **labels** for either rows or columns.

### The Axes

The index and columns are known collectively as the **axes** and each represent a single **axis** of the two-dimensional DataFrame. pandas uses the integer **0** for the index and <strong>1</strong> for the columns.

[1]: images/df_components.png

## What type of object is `bikes`?

Let's verify that `bikes` is indeed a DataFrame with the `type` function.

In [ ]:
type(bikes)

pandas.core.frame.DataFrame

### Fully-qualified name

The above output is something called the **fully-qualified name**. Only the word after the last dot is the name of the type. We have now verified that the `bikes` variable has type `DataFrame`. 

The fully-qualified name always returns the package and module name of where the type was defined. The package name is the first part of the fully-qualified name and, in this case, is `pandas`. The module name is the word immediately preceding the name of the type. Here, it is `frame`.

### Package vs Module

A Python **package** is a directory containing other directories or modules that contain Python code. A Python **module** is a file (typically a text file ending in .py) that contains Python code. 

### Sub-packages

Any directory containing other directories or modules within a Python package is considered a **sub-package**. In this case, `core` is the sub-package.

### Where are the packages located on my machine?

Third-party packages are installed in the `site-packages` directory which itself is set up during Python installation. We can get the actual location with help from the standard library's `site` module's `getsitepackages` function.

In [ ]:
import site
site.getsitepackages()

['/usr/local/lib/python3.6/dist-packages',
 '/usr/lib/python3/dist-packages',
 '/usr/lib/python3.6/dist-packages']

If you navigate to this directory in your file system, you'll find the 'pandas' directory. Within it will be a 'core' directory which will contain the 'frame.py' file. It is this file which contains Python code where the DataFrame class is defined.

## Select a single column from a DataFrame - a Series

To select a single column from a DataFrame, append a set of square brackets, `[]`, to the end of the DataFrame variable name. Place the column name within those brackets to select it. This returns a single column of data as a pandas **Series**. This is a separate (but similar) type of object than a DataFrame.

Let's select the column name `tripduration`, assign it to a variable name, and output the first few values to the screen. The `head` and `tail` methods work the same as they do with DataFrames.

In [ ]:
td = bikes['tripduration']
td.head()

0     993
1     623
2    1040
3     667
4     130
Name: tripduration, dtype: int64

Select the last three values in the Series by passing the `tail` method the integer 3.

In [ ]:
td.tail(3)

50086    824
50087    178
50088    214
Name: tripduration, dtype: int64

## Components of a Series

A Series is a similar type of object as a DataFrame but only contains a single dimension of data. It has two components - the **index** and the **data**. Let's take a look at a stylized Series graphic.

![][0]

It's important to note that a Series has no rows and no columns. In appearance, it resembles a one-column DataFrame, but it technically has no columns. It just has a sequence of values that are labeled by an index.

### The index

A Series index serves as labels for the values. A single **label** or **name** always references a single value. In the above image, the index label **3** corresponds with the value 667. The Series index is virtually identical to the DataFrame index, so the same rules apply to it. Index values can be duplicated and can be types other than integers, such as strings. 

### Output of Series vs DataFrame

Notice that there is no nice HTML styling for the Series. It's just plain text. Below the Series display, you will see a few other items printed to the screen - the **name**, **length**, and **dtype**. These other items are NOT part of the Series itself and are just extra pieces of information to help you understand the Series.

* The **name** is not important right now. If the Series was formed from a column of a DataFrame, it will be set to that column name.
* The **length** is the number of values in the Series
* The **dtype** is the data type of the Series, which will be discussed in an upcoming chapter.

[0]: images/series_components.png

## Changing display options

pandas gives you the ability to change how the output on your screen is displayed. For instance, the default number of columns displayed for a DataFrame is 20, meaning that if your DataFrame has more than 20 columns then only the first and last 10 columns will be shown on the screen. All the other columns will be hidden and unable to be displayed. This is problematic as many DataFrames have more than 20 columns.

### Get current option value with `get_option`

There are a few dozen display options you can control to change the visual representation of your DataFrame. It is not necessary to remember the option names. The official documentation also provides descriptions for all [available options][1]. 

Let's first learn how to retrieve each option value with the `get_option` function. This is not a DataFrame method, but instead a function that is accessed directly from `pd`.  Below are three of the most common options to change.

[1]: http://pandas.pydata.org/pandas-docs/stable/options.html#available-options

In [ ]:
pd.get_option('display.max_columns')

0

In [ ]:
pd.get_option('display.max_rows')

60

In [ ]:
pd.get_option('display.max_colwidth')

50

### Use the `set_option` function to change an option value

To change the option value, use the `set_option` function. You can set as many options as you would like at one time. It's usage is a bit strange. Pass it the option name as a string and follow it immediately with the value you want to set it to. Continue this pattern of option name followed by new value to set as many options as you desire. Below, we set the maximum number of columns to 100 and the maximum number of rows to 4.

In [ ]:
pd.set_option('display.max_columns', 100, 'display.max_rows', 4)

We now read in the housing dataset which contains 81 columns, all of which will be visible.

In [ ]:
housing = pd.read_csv('Pandas/housing.csv')
housing.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0    1          60       RL         65.0     8450   Pave   NaN      Reg   
1    2          20       RL         80.0     9600   Pave   NaN      Reg   
..  ..         ...      ...          ...      ...    ...   ...      ...   
3    4          70       RL         60.0     9550   Pave   NaN      IR1   
4    5          60       RL         84.0    14260   Pave   NaN      IR1   

   LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0          Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1          Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
..         ...       ...       ...       ...          ...        ...   
3          Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4          Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

   Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0        Norm     1Fam     2Story            7            5       2003   
1        Norm     1Fam     1Story            6            8       1976   
..        ...      ...        ...          ...          ...        ...   
3        Norm     1Fam     2Story            7            5       1915   
4        Norm     1Fam     2Story            8            5       2000   

    YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0           2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1           1976     Gable  CompShg     MetalSd     MetalSd       None   
..           ...       ...      ...         ...         ...        ...   
3           1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4           2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

    MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0        196.0        Gd        TA      PConc       Gd       TA           No   
1          0.0        TA        TA     CBlock       Gd       TA           Gd   
..         ...       ...       ...        ...      ...      ...          ...   
3          0.0        TA        TA     BrkTil       TA       Gd           No   
4        350.0        Gd        TA      PConc       Gd       TA           Av   

   BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0           GLQ         706          Unf           0        150          856   
1           ALQ         978          Unf           0        284         1262   
..          ...         ...          ...         ...        ...          ...   
3           ALQ         216          Unf           0        540          756   
4           GLQ         655          Unf           0        490         1145   

   Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0     GasA        Ex          Y      SBrkr       856       854             0   
1     GasA        Ex          Y      SBrkr      1262         0             0   
..     ...       ...        ...        ...       ...       ...           ...   
3     GasA        Gd          Y      SBrkr       961       756             0   
4     GasA        Ex          Y      SBrkr      1145      1053             0   

    GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0        1710             1             0         2         1             3   
1        1262             0             1         2         0             3   
..        ...           ...           ...       ...       ...           ...   
3        1717             1             0         1         0             3   
4        2198             1             0         2         1             4   

    KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0              1          Gd             8        Typ           0         NaN   
1              1          TA             6        Typ           1          TA   
..           ...         ...           ...        ...         ...    

## Exercises

Use the `bikes` DataFrame for the following exercises.

### Exercise 1
<span  style="color:green; font-size:16px">Select the column `events`, the type of weather that was recorded, and assign it to a variable with the same name. Output the first 10 values of it.</span>

### Exercise 2
<span  style="color:green; font-size:16px">What type of object is `events`?</span>

### Exercise 3
<span  style="color:green; font-size:16px">Select the last two rows of the `bikes` DataFrame and assign it to the variable `bikes_last_2`. What type of object is `bikes_last_2`?</span>

### Exercise 4
<span  style="color:green; font-size:16px">Use `pd.reset_option('all')` to reset the options to their default values. Test that this worked. </span>